File: sdk/go/ratelimit.go
package aoa

import (
	"math"
	"net/http"
	"strconv"
	"strings"
	"time"
)

// RateLimit is parsed from response headers. A nil field means the header was
// absent. AOA sends RateLimit-Policy ("default";q=60;w=60), RateLimit
// ("default";r=59;t=60), X-RateLimit-Limit / X-RateLimit-Remaining, and
// Retry-After with 429.
type RateLimit struct {
	// Limit is requests per window (q or X-RateLimit-Limit).
	Limit *int
	// Remaining is requests left in the window (r or X-RateLimit-Remaining).
	Remaining *int
	// ResetSeconds is the upper bound until the quota refills (t).
	ResetSeconds *int
	// WindowSeconds is the window length (w).
	WindowSeconds *int
	// RetryAfterSeconds is Retry-After: the wait before retrying a 429, or the
	// polling interval of a 202 from an asynchronous checkout.
	RetryAfterSeconds *int
}

func parseInt(value string) *int {
	value = strings.TrimSpace(value)
	if value == "" {
		return nil
	}
	parsed, err := strconv.ParseFloat(value, 64)
	if err != nil {
		return nil
	}
	n := int(parsed)
	return &n
}

func structuredParams(value string) map[string]string {
	params := map[string]string{}
	parts := strings.Split(value, ";")
	for _, part := range parts[1:] {
		key, raw, found := strings.Cut(part, "=")
		if found {
			params[strings.TrimSpace(key)] = strings.TrimSpace(raw)
		}
	}
	return params
}

func parseRetryAfter(value string, now time.Time) *int {
	if value == "" {
		return nil
	}
	if seconds := parseInt(value); seconds != nil {
		if *seconds < 0 {
			zero := 0
			return &zero
		}
		return seconds
	}
	moment, err := http.ParseTime(value)
	if err != nil {
		return nil
	}
	seconds := int(math.Max(0, math.Ceil(moment.Sub(now).Seconds())))
	return &seconds
}

func firstNonNil(values ...*int) *int {
	for _, value := range values {
		if value != nil {
			return value
		}
	}
	return nil
}

func parseRateLimit(header http.Header, now time.Time) *RateLimit {
	policy := structuredParams(header.Get("RateLimit-Policy"))
	state := structuredParams(header.Get("RateLimit"))
	info := &RateLimit{
		Limit:             firstNonNil(parseInt(policy["q"]), parseInt(header.Get("X-RateLimit-Limit"))),
		Remaining:         firstNonNil(parseInt(state["r"]), parseInt(header.Get("X-RateLimit-Remaining"))),
		ResetSeconds:      parseInt(state["t"]),
		WindowSeconds:     parseInt(policy["w"]),
		RetryAfterSeconds: parseRetryAfter(header.Get("Retry-After"), now),
	}
	if info.Limit == nil && info.Remaining == nil && info.ResetSeconds == nil &&
		info.WindowSeconds == nil && info.RetryAfterSeconds == nil {
		return nil
	}
	return info
}

File: sdk/typescript/src/rateLimit.ts
/**
 * Rate-limit state parsed from response headers.
 *
 * AOA sends two equivalent sets: the IETF draft pair `RateLimit-Policy`
 * (`"default";q=60;w=60`) and `RateLimit` (`"default";r=59;t=60`), plus the
 * historical `X-RateLimit-Limit` / `X-RateLimit-Remaining`. `Retry-After`
 * comes with 429 (and as a polling interval with a 202).
 */
export interface RateLimitInfo {
  /** Requests allowed per window (`q` or `X-RateLimit-Limit`). */
  limit?: number;
  /** Requests left in the current window (`r` or `X-RateLimit-Remaining`). */
  remaining?: number;
  /** Upper bound, in seconds, until the quota refills (`t`). */
  resetSeconds?: number;
  /** Window length in seconds (`w`). */
  windowSeconds?: number;
  /**
   * `Retry-After` in seconds: the wait before retrying a 429, or the polling
   * interval of a 202 from an asynchronous checkout.
   */
  retryAfterSeconds?: number;
}

type HeaderSource = Pick<Headers, 'get'>;

const toNumber = (value: string | undefined | null): number | undefined => {
  if (value === undefined || value === null || value.trim() === '') {
    return undefined;
  }
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : undefined;
};

/** Parameters of a structured-field item: `"default";q=60;w=60` → {q, w}. */
const structuredParams = (value: string | null): Record<string, string> => {
  const params: Record<string, string> = {};
  if (!value) return params;
  for (const part of value.split(';').slice(1)) {
    const [key, raw] = part.split('=');
    if (key && raw !== undefined) params[key.trim()] = raw.trim();
  }
  return params;
};

/** `Retry-After` is either delta-seconds or an HTTP date. */
const parseRetryAfter = (
  value: string | null,
  now: number,
): number | undefined => {
  if (!value) return undefined;
  const seconds = toNumber(value);
  if (seconds !== undefined) return Math.max(0, seconds);
  const date = Date.parse(value);
  if (Number.isNaN(date)) return undefined;
  return Math.max(0, Math.ceil((date - now) / 1000));
};

/** Returns `null` when the response carries no rate-limit headers at all. */
export const parseRateLimit = (
  headers: HeaderSource,
  now: number = Date.now(),
): RateLimitInfo | null => {
  const policy = structuredParams(headers.get('ratelimit-policy'));
  const state = structuredParams(headers.get('ratelimit'));

  const info: RateLimitInfo = {
    limit: toNumber(policy.q) ?? toNumber(headers.get('x-ratelimit-limit')),
    remaining:
      toNumber(state.r) ?? toNumber(headers.get('x-ratelimit-remaining')),
    resetSeconds: toNumber(state.t),
    windowSeconds: toNumber(policy.w),
    retryAfterSeconds: parseRetryAfter(headers.get('retry-after'), now),
  };

  const present = Object.values(info).some((value) => value !== undefined);
  if (!present) return null;
  for (const key of Object.keys(info) as (keyof RateLimitInfo)[]) {
    if (info[key] === undefined) delete info[key];
  }
  return info;
};
